lib/virtual-fs/src/lib.rs

```diff
@@ -0,0 +1,319 @@
+    fn last_accessed(&self) -> u64;
+
+    /// the last time the file was modified in nanoseconds as a UNIX timestamp
+    fn last_modified(&self) -> u64;
+
+    /// the time at which the file was created in nanoseconds as a UNIX timestamp
+    fn created_time(&self) -> u64;
+
+    /// set the last time the file was accessed in nanoseconds as a UNIX timestamp
+    fn set_last_accessed(&self, _last_accessed: u64) {
+        debug!("{:?} did nothing in VirtualFile::set_last_accessed due to using the default implementation", self);
+    }
+
+    /// set the last time the file was modified in nanoseconds as a UNIX timestamp
+    fn set_last_modified(&self, _last_modified: u64) {
+        debug!("{:?} did nothing in VirtualFile::set_last_modified due to using the default implementation", self);
+    }
+
+    /// set the time at which the file was created in nanoseconds as a UNIX timestamp
+    fn set_created_time(&self, _created_time: u64) {
+        debug!(
+            "{:?} did nothing in VirtualFile::set_created_time to using the default implementation",
+            self
+        );
+    }
+
+    /// the size of the file in bytes
+    fn size(&self) -> u64;
+
+    /// Change the size of the file, if the `new_size` is greater than the current size
+    /// the extra bytes will be allocated and zeroed
+    fn set_len(&mut self, new_size: u64) -> Result<(), FsError>;
+
+    /// Request deletion of the file
+    fn unlink(&mut self) -> Result<(), FsError>;
+
+    /// Store file contents and metadata to disk
+    /// Default implementation returns `Ok(())`.  You should implement this method if you care
+    /// about flushing your cache to permanent storage
+    fn sync_to_disk(&self) -> Result<(), FsError> {
+        Ok(())
+    }
+
+    /// Moves the file to a new location
+    /// NOTE: the signature of this function will change before stabilization
+    // TODO: stablizie this in 0.7.0 or 0.8.0 by removing default impl
+    fn rename_file(&self, _new_name: &std::path::Path) -> Result<(), FsError> {
+        panic!("Default implementation for now as this method is unstable; this default implementation or this entire method may be removed in a future release.");
+    }
+
+    /// Returns the number of bytes available.  This function must not block
+    fn bytes_available(&self) -> Result<usize, FsError>;
+
+    /// Used for polling.  Default returns `None` because this method cannot be implemented for most types
+    /// Returns the underlying host fd
+    fn get_raw_fd(&self) -> Option<i32> {
+        None
+    }
+}
+
+// Implementation of `Upcastable` taken from https://users.rust-lang.org/t/why-does-downcasting-not-work-for-subtraits/33286/7 .
+/// Trait needed to get downcasting from `VirtualFile` to work.
+pub trait Upcastable {
+    fn upcast_any_ref(&'_ self) -> &'_ dyn Any;
+    fn upcast_any_mut(&'_ mut self) -> &'_ mut dyn Any;
+    fn upcast_any_box(self: Box<Self>) -> Box<dyn Any>;
```

Comment:
This method is not used anywhere

Reply:
It's a public method, we can review its utility but it's useful when working with `dyn Thing` and you might, as a user, care about getting your concrete thing out of the dynamic thing. This is a useful part of the `WasiFile` API which allows users to define whatever logic they want for `WasiFile`s